cmd/kubeadm/app/cmd/phases/certs.go

```diff
@@ -17,87 +17,360 @@ limitations under the License.
-func RunSelfSign(config *kubeadmapi.MasterConfiguration) error {
-	if err := validateArgs(config); err != nil {
-		return fmt.Errorf("The argument validation failed: %v", err)
+// CreatePKIAssets will create and write to disk all PKI assets necessary to establish the control plane.
+// Please note that this action is a bulk action calling all the atomic certphase actions
+func CreatePKIAssets(cfg *kubeadmapi.MasterConfiguration) error {
+
+	certActions := []func(cfg *kubeadmapi.MasterConfiguration) error{
+		createOrUseCACertAndKey,
+		createOrUseAPIServerCertAndKey,
+		createOrUseAPIServerKubeletClientCertAndKey,
+		createOrUseServiceAccountKeyAndPublicKey,
+		createOrUseFrontProxyCACertAndKey,
+		createOrUseFrontProxyClientCertAndKey,
 	}
 
-	// If it's possible to detect the default IP, add it to the SANs as well. Otherwise, just go with the provided ones
-	ip, err := netutil.ChooseBindAddress(net.ParseIP(config.API.AdvertiseAddress))
-	if err == nil {
-		config.API.AdvertiseAddress = ip.String()
+	for _, action := range certActions {
+		err := action(cfg)
+		if err != nil {
+			return err
+		}
 	}
 
-	err = certphase.CreatePKIAssets(config)
-	if err != nil {
-		return err
+	fmt.Printf("[certificates] Valid certificates and keys now exist in %q\n", cfg.CertificatesDir)
+
+	return nil
+}
+
+// createOrUseCACertAndKey create a new self signed CA, or use the existing one.
+func createOrUseCACertAndKey(cfg *kubeadmapi.MasterConfiguration) error {
+
+	return createOrUseCertificateAuthorithy(
+		cfg.CertificatesDir,
+		kubeadmconstants.CACertAndKeyBaseName,
+		"CA",
+		certphase.NewCACertAndKey,
+	)
+}
+
+// createOrUseAPIServerCertAndKey create a new CA certificate for apiserver, or use the existing one.
+// It assumes the CA certificates should exists into the CertificatesDir
+func createOrUseAPIServerCertAndKey(cfg *kubeadmapi.MasterConfiguration) error {
+
+	return createOrUseSignedCertificate(
+		cfg.CertificatesDir,
+		kubeadmconstants.CACertAndKeyBaseName,
+		kubeadmconstants.APIServerCertAndKeyBaseName,
+		"API server",
+		func(caCert *x509.Certificate, caKey *rsa.PrivateKey) (*x509.Certificate, *rsa.PrivateKey, error) {
+			return certphase.NewAPIServerCertAndKey(cfg, caCert, caKey)
+		},
+	)
+}
+
+// create a new CA certificate for kubelets calling apiserver, or use the existing one
+// It assumes the CA certificates should exists into the CertificatesDir
+func createOrUseAPIServerKubeletClientCertAndKey(cfg *kubeadmapi.MasterConfiguration) error {
+
+	return createOrUseSignedCertificate(
+		cfg.CertificatesDir,
+		kubeadmconstants.CACertAndKeyBaseName,
+		kubeadmconstants.APIServerKubeletClientCertAndKeyBaseName,
+		"API server kubelet client",
+		certphase.NewAPIServerKubeletClientCertAndKey,
+	)
+}
+
+// createOrUseServiceAccountKeyAndPublicKey create a new public/private key pairs for signing service account user, or use the existing one.
+func createOrUseServiceAccountKeyAndPublicKey(cfg *kubeadmapi.MasterConfiguration) error {
```

Comment:
`NewServiceAccountSigningKey` is the only method that creates a key and extract public key file, while other actions create key and certificate pairs.

However I agree with you that maintaining the same pattern also in this case could help in readability, so Done!